test/src/unit-arrow.cc

```diff
@@ -0,0 +1,371 @@
+/**
+ * @file   unit-arrow_io.cc
+ *
+ * @section LICENSE
+ *
+ * The MIT License
+ *
+ * @copyright Copyright (c) 2017-2020 TileDB Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * @section DESCRIPTION
+ *
+ * Tests for the TileDB Arrow integration.
+ */
+
+#include "catch.hpp"
+#include "helpers.h"
+#include "tiledb/sm/cpp_api/arrowio"
+#include "tiledb/sm/cpp_api/tiledb"
+#include "tiledb/sm/misc/constants.h"
+#include "tiledb/sm/misc/uri.h"
+#include "tiledb/sm/misc/utils.h"
+
+#include <pybind11/embed.h>
+#include <pybind11/pybind11.h>
+
+using namespace tiledb;
+namespace py = pybind11;
+
+namespace {
+struct CPPArrayFx {
+ public:
+  CPPArrayFx(std::string uri, const uint64_t col_size = 111)
+      : vfs(ctx)
+      , uri_(uri) {
+    using namespace tiledb;
+
+    if (vfs.is_dir(uri))
+      vfs.remove_dir(uri);
+
+    const uint64_t d1_tile = 10;
+
+    Domain domain(ctx);
+    auto d1 =
+        Dimension::create<int>(ctx, "d1", {{0, (int)col_size - 1}}, d1_tile);
+    domain.add_dimensions(d1);
+
+    std::vector<Attribute> attrs;
+    attrs.insert(
+        attrs.end(),
+        {Attribute::create<int8_t>(ctx, "int8"),
+         Attribute::create<int16_t>(ctx, "int16"),
+         Attribute::create<int32_t>(ctx, "int32"),
+         Attribute::create<int64_t>(ctx, "int64"),
+
+         Attribute::create<uint8_t>(ctx, "uint8"),
+         Attribute::create<uint16_t>(ctx, "uint16"),
+         Attribute::create<uint32_t>(ctx, "uint32"),
+         Attribute::create<uint64_t>(ctx, "uint64"),
+
+         Attribute::create<float>(ctx, "float32"),
+         Attribute::create<double>(ctx, "float64")});
+
+    // must be constructed manually to get TILEDB_STRING_UTF8 type
```

Comment:
We don't have a type mapping to `TILEDB_STRING_UTF8`, and I'm not sure whether we even _should_ because `std::string` is not guaranteed to be UTF8.

Likewise for `DATETIME_NS`: I don't know what C++ type would map to that in the templated constructor.

Reply:
There's probably some template magic we could do to allow writing
```
Attribute::create<TILEDB_DATETIME_NS>(ctx, "datetime_ns");
```
which would be nice!

Reply:
Got it, thanks!